pkg/reconciler/taskrun/taskrun.go

```diff
@@ -191,6 +193,26 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	return multierror.Append(merr, c.updateStatusLabelsAndAnnotations(tr, original)).ErrorOrNil()
 }
 
+// Run any async logic that may be required at start-up time. This method is used
+// to emit events, notifications or any other async operation
+func (c *Reconciler) preRunAsyncHook(ctx context.Context, tr *v1alpha1.TaskRun) {
+	c.Logger.Infof("preRunAsyncHook: %s", tr.Name)
+
+	// Emit event
+	afterCondition := tr.Status.GetCondition(apis.ConditionSucceeded)
+	reconciler.EmitEvent(c.Recorder, nil, afterCondition, tr)
+}
+
+// Run any async logic that may be required once the tr is successfully reconciled
+// This method is used to emit events, notifications or any other async operation
+func (c *Reconciler) postRunAsyncHook(ctx context.Context, tr *v1alpha1.TaskRun, beforeCondition *apis.Condition) {
+	c.Logger.Infof("postRunAsyncHook: %s", tr.Name)
+
+	// Emit event
+	afterCondition := tr.Status.GetCondition(apis.ConditionSucceeded)
+	reconciler.EmitEvent(c.Recorder, beforeCondition, afterCondition, tr)
+}
```

Comment:
do these need to be functions on the reconciler? maybe we could make a new struct and initialize it with the function to emit events?

Reply:
I put them on the reconciler to give them access to anything that they may need from it, but indeed I was thinking of changing the implementation there.

Reply:
I decided to remove these functions - at least until I actually need to use them.
Emitting events is already non-blocking - down in the implementation of `generateEvent`.
So there was little value in adding functions here.
Besides, fetching the afterCondition in a goroutine triggered a data race in the unit tests, which is a good sign that I should not do that.